File: WebAPI/Models/Product.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Collections.Generic;

namespace WebAPI.Models
{
    public record Product
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int Id { get; set; }

        [Required]
        [Display(Name = "Product name")]
        [MaxLength(100, ErrorMessage = "Maximum length for product name is 100 characters")]
        public string Name { get; set; }

        [Required]
        [Range(1, int.MaxValue, ErrorMessage = "Price must be greater than 0")]
        [RegularExpression(@"^\d*\.?\d{0,2}$", ErrorMessage = "Price can't have more than 2 decimal places")]
        public double Price { get; set; }
        // [Required]
        [Range(0, int.MaxValue, ErrorMessage = "Ammount must be greater than 0")]
        public double Ammount { get; set; }

        public double Discount { get; set; }

        public string UnitOfMeasurement { get; set; }

        [MaxLength(250, ErrorMessage = "Maximum length for product descitption is 250 characters")]
        public string Description { get; set; }

        [Required]
        public int? ProductCategoryId { get; set; }

        [ForeignKey("ProductCategoryId")]
        public ProductCategory ProductCategory { get; set; }

        public string ImageName { get; set; }

        public double DiscountedPrice { get; private set; }

        public List<ProductReview> ProductReviews { get; set; }

        public void CalculateDiscountedPrice()
        {
            DiscountedPrice = Price - Math.Round(Price * (Discount / 100), 2);
        }
    }
}

File: Angular_UI/src/app/models/product.ts
import { ISearchable } from "../interfaces/ISearchable";
import { productCategory } from "./productCategory";

export class Product implements ISearchable {

    id: number;
    name: string;
    price: number;
    rating: number;
    productCategoryId: number;
    productCategory: productCategory;
    discountedPrice: number;
    ammount: number;
    description: string;
    imageName: string;

    searchProperty: string;

    constructor(id: number, name: string, price: number, 
        rating: number, productCategoryId: number, productCategory: productCategory,
        discountedPrice: number, ammount: number, description: string, imageName: string) {
    // constructor(id: number, name: string, price: number, rating: number, category: string, imageName: string) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.rating = rating;
        this.productCategoryId = productCategoryId;
        this.productCategory = productCategory;
        this.discountedPrice = discountedPrice;
        this.ammount = ammount;
        this.description = description;
        this.imageName = imageName;
        // this.category = category;
    }
}

// export function doesProductPriceBelongToPriceRange (fromPrice: number, toPrice: number) {
//     return ((toPrice == 0) || (this.price >= fromPrice && this.price <= toPrice));
//   }